src/db/roomstore.ts

```diff
@@ -0,0 +1,365 @@
+
+    constructor(private db: IDatabaseConnector) {
+        this.entriesMatrixIdCache = new Map();
+    }
+
+    public async upsertEntry(entry: IRoomStoreEntry) {
+        const promises: Promise<void>[] = [];
+
+        const row = (await this.db.Get("SELECT * FROM room_entries WHERE id = $id", {id: entry.id})) || {};
+
+        if (!row.id) {
+            // Doesn't exist at all, create the room_entries row.
+            const values = {
+                id: entry.id,
+                matrix: entry.matrix ? entry.matrix.roomId : null,
+                remote: entry.remote ? entry.remote.roomId : null,
+            };
+            try {
+                await this.db.Run(`INSERT INTO room_entries VALUES ($id, $matrix, $remote)`, values);
+                log.verbose("Created new entry " + entry.id);
+            } catch (ex) {
+                log.error("Failed to insert room entry", ex);
+                throw Error("Failed to insert room entry");
+            }
+        }
+
+        const matrixId = entry.matrix ? entry.matrix.roomId : null;
+        const remoteId = entry.remote ? entry.remote.roomId : null;
+        const mxIdDifferent = matrixId !== row.matrix_id;
+        const rmIdDifferent = remoteId !== row.remote_id;
+        // Did the room ids change?
+        if (mxIdDifferent || rmIdDifferent) {
+            if (matrixId) {
+                this.entriesMatrixIdCache.delete(matrixId);
+            }
+            const items: string[] = [];
+
+            if (mxIdDifferent) {
+                items.push("matrix_id = $matrixId");
+            }
+
+            if (rmIdDifferent) {
+                items.push("remote_id = $remoteId");
+            }
+
+            await this.db.Run(`UPDATE room_entries SET ${items.join(", ")} WHERE id = $id`,
+                {
+                    id: entry.id,
+                    matrixId: matrixId as string|null,
+                    remoteId: remoteId as string|null,
+                },
+            );
+        }
+
+        // Matrix room doesn't store any data.
+        if (entry.remote) {
+            await this.upsertRoom(entry.remote);
+        }
+    }
+
+    public async getEntriesByMatrixId(matrixId: string): Promise<IRoomStoreEntry[]> {
+        const cached = this.entriesMatrixIdCache.get(matrixId);
+        if (cached && cached.ts + ENTRY_CACHE_LIMETIME > Date.now()) {
+            return cached.e;
+        }
+        const entries = await this.db.All(
+            "SELECT * FROM room_entries WHERE matrix_id = $id", {id: matrixId},
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            let remote: RemoteStoreRoom|null = null;
+            if (entry.remote_id) {
+                const remoteId = entry.remote_id as string;
+                const row = await this.db.Get(
+                    "SELECT * FROM remote_room_data WHERE room_id = $remoteId",
+                    {remoteId},
+                );
+                if (row) {
```

Comment:
actually, thinking about it more it might be a good idea to remove the entries without a remote. That should kill #225 once and for all as it *should* always be present